### ISSUE TITLE
Error on sync/async subscribe to pid(s)/atom(s) returned from into_stages

### ISSUE THREAD

**Issue by zeeshanlakhani** [elixir-lang/flow#12]
So, we wanted to have a way to run `GenStage.sync/async.subscribe` and subscribe_to an `into_stages` producer_consumer pid that was already running, which should work, right, as per docs?

Here's some sample code that returns an error:

* First, let's just use a consumer from the general docs

```elixir
defmodule C do
  use GenStage

  def start_link() do
    GenStage.start_link(C, :ok)
  end

  def init(:ok) do
    {:consumer, :the_state_does_not_matter}
  end

  def handle_events(events, _from, state) do
    # Wait for a second.
    :timer.sleep(1000)

    # Inspect the events.
    IO.inspect(events)

    # We are a consumer, so we would never emit items.
    {:noreply, [], state}
  end
end
```

* now, here's the the flow which gives us a {:ok, proc}, when ending w/ into_stages

```elixir
{:ok, flowpid} = Flow.from_enumerable(1..20) |> Flow.filter(& rem(2, &1) == 0) |> Flow.into_stages([])

```

* now, here's starting the consumer and subscribing (after the fact, as the flow already exists)

```elixir
{:ok, c} = C.start_link() 

GenStage.async_subscribe(c, to: flowpid)
```

* then, here comes the error

```bash
** (EXIT from #PID<0.429.0>) an exception was raised:
    ** (FunctionClauseError) no function clause matching in Flow.Coordinator.handle_info/2
        (flow) lib/flow/coordinator.ex:69: Flow.Coordinator.handle_info({:"$gen_producer", {#PID<0.457.0>, #Reference<0.0.1.213>}, {:subscribe, nil, []}}, %{intermediary: [{#PID<0.462.0>, []}, {#PID<0.463.0>, []}, {#PID<0.464.0>, []}, {#PID<0.465.0>, []}, {#PID<0.466.0>, []}, {#PID<0.467.0>, []}, {#PID<0.468.0>, []}, {#PID<0.469.0>, []}], parent_ref: #Reference<0.0.1.200>, producers: [#PID<0.461.0>], refs: [#Reference<0.0.1.192>, #Reference<0.0.1.193>, #Reference<0.0.1.194>, #Reference<0.0.1.195>, #Reference<0.0.1.196>, #Reference<0.0.1.197>, #Reference<0.0.1.198>, #Reference<0.0.1.199>], supervisor: #PID<0.460.0>})
        (stdlib) gen_server.erl:601: :gen_server.try_dispatch/4
        (stdlib) gen_server.erl:667: :gen_server.handle_msg/5
        (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
```

I came up with a workable fix, but I wanted to see a) if it's even worthwhile and b) how the coordinator matching/checking can/should be updated to allow for sync or **async** subscriptions, as for any *initial* consumers, everything is a sync_subscribe as per https://github.com/elixir-lang/flow/blob/master/lib/flow/coordinator.ex#L54. Maybe we always want it sync no matter what? As of now, it's seems difficult to allow for async here, unless I'm missing something. Additionally, using a handle_info for this may just be a bad hack. So, I hope for a good solution :). 

The hack:

```diff
diff --git a/lib/flow/coordinator.ex b/lib/flow/coordinator.ex
index ece4416..0da06bd 100644
--- a/lib/flow/coordinator.ex
+++ b/lib/flow/coordinator.ex
@@ -66,6 +66,13 @@ defmodule Flow.Coordinator do
     {:noreply, state}
   end

+  def handle_info({:"$gen_producer", {consumer, _}, _}, %{intermediary: intermediary} = state) do
+    for {pid, _} <- intermediary do
+        subscribe(consumer, pid)
+    end
+    {:noreply, state}
+  end
+
   def handle_info({:DOWN, ref, _, _, reason}, %{parent_ref: ref} = state) do
     {:stop, reason, state}
   end
```

Thanks.

**Comment by josevalim** [comment]
@zeeshanlakhani although the error message is horrible, the limitation exists on purpose. I believe there are inherent race conditions in joining streams after they started.

For example, if the producer finished and you subscribe after it finished, you will no longer receive a notification that it is done. So you can get into a deadlock where the late consumer never exits, because it was never told the stream is done, and the producer never exits because the consumer is there.

The solution may be simple: if we subscribe to a producer that is effectively done, we should send a notification right after.

If the race condition is solved, then we can probably gladly perform the redirection mechanism you proposed.

**Comment by zeeshanlakhani** [comment]
@josevalim Yeah, I totally get that. We'd really like to do this as a way to piece together varying prod-consumer/consumer *subflows* (into-stages -> from-stages), after the fact, for possibly long-lived flows. 

I agree w/ the need to handle-on *done* producers if a producer is completed before the consumer subscribes to it. I'm guessing that, w/ async_notify as is being called, we don't need new code to handle consumer subscriptions before a *done*, since that seems covered, right?

I'll try and see how I can solve the race condition and if it's simple enough. If you have more thoughts or get to something first, let me know :).

**Comment by josevalim** [comment]
> I agree w/ the need to handle-on done producers if a producer is completed before the consumer subscribes to it. I'm guessing that, w/ async_notify as is being called, we don't need new code to handle consumer subscriptions before a done, since that seems covered, right?

Yes, exactly. It should be a matter of sending the notification message on handle_subscribe to the consumer if the current stage is "down".

**Comment by zeeshanlakhani** [comment]
Also, to the question of sync/async, should we assume that any subscribe to a prod_consumer this way should be sync (even when using async_subscribe)? My redirection mechanism, as is, is just `sync` specific via that subscribe function, as I'm not sure how to *know* it's coming from an `async` call.

**Comment by josevalim** [comment]
It seems what you are doing is enough and it should work for both sync and async. The sync and async is about the communication between the client and consumer and here we are focused on the consumer messaging the producer.

What do you think?

**Comment by zeeshanlakhani** [comment]
That is a good point, i.e. sync/async is regarding client and consumer. Makes sense with that thought @josevalim.

**Comment by josevalim** [comment]
Fixed in #19.
